### PR TITLE
Cassandra metadata cache

### PR DESCRIPTION
**Pull request by aravindanr**
Pull Request type
----
- [ ] Bugfix
- [ ] Feature
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes (Please run `./gradlew generateLock saveLock` to refresh dependencies)
- [ ] WHOSUSING.md
- [ ] Other (please describe):

**NOTE**: Please remember to run `./gradlew spotlessApply` to fix any format violations.

Changes in this PR
----

_Caches for CassandraMetadataDAO and CassandraEventHandlerDAO are extracted into decorators, CacheableMetadataDAO and CacheableEventHandlerDAO respectively_

Issue: Including cache within the DAO makes it impossible to opt-out of it. 

Alternatives considered
----

_None_

### CODE REVIEW

File: cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/config/cache/CacheableEventHandlerDAO.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            cassandraEventHandlerDAO.removeEventHandler(name);
+        } finally {
+            eventHandlerCache.remove(name);
+        }
+        ;
+    }
+
+    @Override
+    public List<EventHandler> getAllEventHandlers() {
+        return cassandraEventHandlerDAO.getAllEventHandlers();
```
Comment (jxu-nflx) [inline cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/config/cache/CacheableEventHandlerDAO.java:90]: Not reading from cache first?

File: cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/config/cache/CacheableEventHandlerDAO.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    @Override
+    public List<EventHandler> getEventHandlersForEvent(String event, boolean activeOnly) {
+        return cassandraEventHandlerDAO.getEventHandlersForEvent(event, activeOnly);
```
Comment (jxu-nflx) [inline cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/config/cache/CacheableEventHandlerDAO.java:95]: read from cache first?

File: cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/dao/CassandraMetadataDAO.java [hunk 3]
```diff
@@ -304,22 +288,6 @@ public List<WorkflowDef> getAllWorkflowDefs() {
...
-            getAllTaskDefsFromDB().forEach(taskDef -> map.put(taskDef.getName(), taskDef));
-            this.taskDefCache = map;
-            LOGGER.debug("Refreshed task defs, total num: " + this.taskDefCache.size());
-        } catch (Exception e) {
-            Monitors.error(CLASS_NAME, "refreshTaskDefs");
-            LOGGER.error("refresh TaskDefs failed ", e);
-        }
-    }
-
     private TaskDef getTaskDefFromDB(String name) {
```
Comment (jxu-nflx) [inline cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/dao/CassandraMetadataDAO.java:291]: Add taskdefname tag to the metric?

File: cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/config/cache/CacheableMetadataDAO.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    public void removeWorkflowDef(String name, Integer version) {
+        cassandraMetadataDAO.removeWorkflowDef(name, version);
+    }
+
+    @Override
+    public List<WorkflowDef> getAllWorkflowDefs() {
+        return cassandraMetadataDAO.getAllWorkflowDefs();
+    }
+
+    private void refreshTaskDefsCache() {
```
[the remark this answers is not in the document]
Comment (jxu-nflx) [reply cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/config/cache/CacheableMetadataDAO.java:136]: If not, we can add a metrics to the `getAllTaskDefsFromDB` method of CassdrandMetaDataDao

[the remark this answers is not in the document]
Comment (aravindanr) [reply cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/config/cache/CacheableMetadataDAO.java:136]: > Does @trace create metrics for this method call as well?

No. @Trace does not work on private methods. The plan is to use Spring Cacheable in the internal implementation which allows greater flexibility in configuration and also exposes metrics.

File: cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/dao/CassandraMetadataDAO.java (hunk 3 above)
Comment (aravindanr) [reply cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/dao/CassandraMetadataDAO.java:291]: added

File: cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/config/cache/CacheableEventHandlerDAO.java (hunk 1 above)
Comment (aravindanr) [reply cassandra-persistence/src/main/java/com/netflix/conductor/cassandra/config/cache/CacheableEventHandlerDAO.java]: Done

**Pull request merged**